lib_eio/flow.ml

```diff
@@ -78,6 +81,9 @@ let buffer_sink b =
           Buffer.add_string b (Cstruct.to_string ~len:got buf)
         done
       with End_of_file -> ()
+
+    method write_from buf =
+      Buffer.add_bytes b (Cstruct.to_bytes buf) (* XXX slow *)
```

Comment:
I don't think there's a faster way to do it given the `Buffer` API. We can just note in the mli that this might not be very efficient (in fact, it already points you in the direction of `Buf_read`).

Reply:
Given that `Buffer` doesn't expose useful functions for that, it might be a good idea to also expose a better implementation of it (`{mutable buf: bytes;  mutable len: int }` basically), if only to get a `string` out of a `Flow.source` easily.

Reply:
@avsm is thinking about changing Cstruct to be backed by `bytes` rather than bigarray. If that happens, `Eio.Buf_read` will work for this case (see https://github.com/ocaml-multicore/eio/issues/140).

Reply:
> @avsm is thinking about changing Cstruct to be backed by `bytes` rather than bigarray. If that happens, `Eio.Buf_read` will work for this case (see #140).

Aye, I checked that, I've been thinking of implementing something like `Mbuf.t` which could be backed by a chain of Bytes or Cstruct, a bit inspired by how mbuf(9) works. With that we can guarantee that the backing Bytes is always > wosize so it goes directly in the major heap, and also, if needed, swap to Cstruct transparently in the future if things change in the GC.

It would be something like "allocate a fixed bytes size, say 256, start writing at say offset 63 so future prepends are free, appending a few words would just write to the existing chunk, appending a longer chunk would link it to the chain". Maybe have a big (>1500) and little ( < 512 ).

We could then make full usage of writing a backing method in Eio to turn that into the iovecs.

I don't think Cstruct or Bytes directly is a good way for people to build packets.

Reply:
Note: we already have `Buf_write`, which is similar (though it doesn't allocate room for prepending). It does allow appending cstructs without copying, which results in using `writev` automatically.